Library/Homebrew/formula.rb

```diff
@@ -1706,14 +1706,49 @@ def self.installed
     end.uniq(&:name)
   end
 
-  # An array of all installed {Formula} without dependents
+  # An array of all installed {Formula} with {Cask} dependents.
   # @private
-  def self.installed_formulae_with_no_dependents(formulae = installed)
+  def self.installed_formulae_with_cask_dependents
+    require "cask/caskroom"
+
+    Cask::Caskroom.casks
+                  .flat_map { |cask| cask.depends_on[:formula] }
+                  .compact
+                  .map { |f| Formula[f] }
+                  .flat_map { |f| [f, *f.runtime_formula_dependencies].compact }
+  end
+
+  # An array of all installed {Formula} without {Formula} dependents
+  # @private
+  def self.installed_formulae_with_no_formula_dependents(formulae = installed)
     return [] if formulae.blank?
 
     formulae - formulae.flat_map(&:runtime_formula_dependencies)
   end
 
+  # Recursive function that returns an array of installed {Formula} without
+  # {Formula} dependents that weren't installed on request.
+  # @private
+  def self.unused_formulae_with_no_formula_dependents(formulae)
```

Comment:
I think we should also remove formulae listed in `HOMEBREW_NO_CLEANUP_FORMULAE` here.

Reply:
Good idea. Would warrant a documentation change, too.

Reply:
Yeah, that makes sense to me too.

Reply:
I ended up adding this to `Cleanup#autoremove` mainly because that check already exists in the `Cleanup` class and I didn't want to require that inside of `Formula`.